### DISCUSSION TITLE
Can we talk about the class sort order? [Prettier Plugin]

### DISCUSSION THREAD

**Discussion by asimpletune** [tailwindlabs/tailwindcss#12804] [Ideas]
Originally I was going to write this in [the PR](https://github.com/biomejs/biome/pull/1362) for biomejs, but then I figured it's actually a tailwind-wide question.

As I understand, the [sort order](https://tailwindcss.com/blog/automatic-class-sorting-with-prettier#how-classes-are-sorted) is based on the order of how class are generated  in the resulting CSS. I think this is a very good start, and handles most of what one wants. However with variants I don't think that is the case, and I'd like to explain why. 

The best possible outcome of this discussion is I just realize  that I'm wrong (extremely likely possibility) and that's 100% ok with me, even preferred. If I'm wrong, then we can all sleep at night knowing we already do things the best we know how. However, there is the possibility that we do not currently have the best way of sorting classes, in which case there is room for improvement. So let's hope I'm wrong, as that is certainly easier, but not shy away from the discussion.

Ok, here goes. 

## The rules behind Tailwind's class ordering

So the plugin's ordering works by the order of the generated CSS. This is mostly very good. 

First, definitely, it is important to have classes that modify/override the same rules be sorted by specificity, so it's clear which takes precedence. The example from the docs of `p-4 pt-2` makes the case crystal clear, since `pt-2` is _more specific_ than `p-4`. The specific case _should_ override the generic case. 

By that same logic, another way of expressing this rule is that **"like goes with like"**, and ties go to whichever is more specific. After all, what could be more alike than two classes modifying the same rule?

Additionally,  in the tailwind docs there is mention of a high level goal for the ordering as well:

> The actual order of the different utilities is loosely based on the box model, and tries to put high impact classes that affect the layout at the beginning and decorative classes at the end, while also trying to keep related utilities together:

This also makes sense, with the most impactful, layout rules coming first, decorative ones coming last.

There's a third goal, which I'm just mentioning for completeness, that user defined classes come first. 

Therefore, according to the docs, we have 3 sorting rules: 

1. user defined classes
2. layout first, then decoration
3. like with like, with the more specific coming later

On all accounts, I am 100% in agreement and I think it all makes perfect sense. However, there is another rule, regarding modifiers that basically conflicts with the first 3! It's a complete contradiction. 

## The issue comes with modifiers 

> Modifiers like hover: and focus: are grouped together and sorted after any plain utilities

The problem is this violates the rules above. It basically, logically, means 'order by impact, ties are broken by specificity, ... but modifiers are immune from all of this and are tacked on in the end'. Why? 

Like I said, please don't mistake my meaning. This is better than nothing, but having the modifiers come at the end is honestly not useful. 

I understand that modifiers need to be distinguished from regular utility classes, that is clear, but it doesn't need to happen with sorting! I know something is a modifier by virtue of it having a modifier in the name. So, from this view, we're duplicating the benefits that we already get, which _on its own_ isn't harmful but it does beg the question:

**Are there actively negative affects of distinguishing modifiers from utility classes by sorting?**

The downside of having it this way is we undo all the good we accomplished with our 3 sorting rules from above, because now I must read all of my styles to understand what they do. By this logic, we gain a benefit that we already had, but it comes at the price of losing other benefits, so by this view it's a net loss. I will try to make this last point more clear, before explaining how this 'net loss' is totally avoidable.

Since modifiers come last, and a modifier can precede any utility class, that implies that I have to read _all_ of my code to understand the layout aspects, because they're no longer sorted earlier. Further more, I can no longer rely on the  "like matches like" rule, because there may be more "alike" that are hidden somewhere in whatever modifiers I have. 

Even worse, this is for every modifier as well too! It's not like there's an invisible line where  on one side is all the non-modifier stuff is sorted, and then on the other the modifier stuff is sorted. It's more like on the modifier side of the line, the 3 rules are respected but for each modifier! This means I MUST read every modifier to make sure nothing important was missed. In other words, I can't scan my code and rely upon the ordering to tell me what I need to know, because things aren't sorted anymore by impact or like with like.

However, this problem is only important if you have a lot of modifiers. Most people probably don't, which is why I want to give a real, non-contrived example of what I think is a reasonable thing to make, that at times has unnecessary friction because of this:

## 'Tabbed Window' example

I made a  Mac-style [tabbed window](https://spenc.es/writing/how-to-create-toggle-menus-with-tailwindcss/#step-1-menu-layout-tabs) for displaying code that's all related in different tabs. I try to make everything I write work without requiring JS to function correctly, and I also optimize for browser reader modes. This means my markup is going to be a little bit more complicated than I'd like, and  I have to therefore use modifiers a lot. The end result is really cool, I think, and I love that in reader mode they're just normal document sections, complete with their tab's title being the section title. Here are some screenshots:

Normal mode:
<img width="825" alt="Screenshot 2024-01-22 at 3 40 22 PM" src="https://github.com/tailwindlabs/tailwindcss/assets/2670813/14a772af-6099-403b-98c8-ee6c80a1172d">

After clicking the green button:
<img width="1404" alt="Screenshot 2024-01-22 at 3 40 49 PM" src="https://github.com/tailwindlabs/tailwindcss/assets/2670813/03dc1e86-0c9a-4a89-8cbd-d2aec550f9af">

After clicking the yellow button:
<img width="892" alt="Screenshot 2024-01-22 at 3 41 04 PM" src="https://github.com/tailwindlabs/tailwindcss/assets/2670813/09a939b6-43f3-4e4d-a0ff-60322f9d98bc">

The end result is it basically works how you expect, but the logic is complicated. 

The most important of that logic comes down to layout, as I want to expand the window if they press the expand button, or shrink if they want to minimize that example. This alone would be fine, because the layout isn't _that_ complicated after all, however I also have to handle the styling of the different tabs. 

When one tab is selected, I use a document fragment link (to that tab) in combination with some tailwind modifiers to make sure the expected thing happens with the tabs's styling. The default tab though has no document fragment link in the browser so it may not use any modifier at all. And things quickly become sort of a mess, as I'm sure you can understand. There's making sure the tabs background is lighter, its text is lighter, but the non-selected ones are respectively darker, and that when you hover over a tab it becomes medium levels of light. 

Having a default that only applies to one tab, plus decoration that happens when one tab is selected, and separate decoration for everything if a tab is not selected, which the default tab must also adopt because it's not the default case any more is... very complicated. 

Ok, so it's complicated, but my point is that the modifier sort order does not help. I have peers, I have groups, I have all sorts of stuff going on there. They all handle layout and decoration, but then those things also exist in the non-modifier area of the class list. I still think this is a cool thing to make, and I think it's not some extreme example of having crazy amounts of modifiers. The code still consists of totally readable, semantic HTML, and I barely had to write any CSS. 

## We already have the better way

I want to thank whoever read all this way, and I don't want to end this without highlighting what I think would be a better way to do this. Therefore, I think the first 3 rules actually are great, and modifiers should just be forced to respect them.

1. user defined classes
2. layout first, then decoration (even if preceded by a modifier)
3. like with like, with the more specific coming later (even if preceded by a modifier)

This actually makes a lot more sense to with how they work, since modifiers will already override competing rules of less specific stuff. I think the difficulty would be that it is just more complicated to implement, but that's a separate discussion. I'd like to know what other users and maintainers think. 

Thanks for reading.

**Comment by wolfspyre** [comment]
I don't have a solution... but I wanted to acknowledge and highlight the thought, and effort that went into articulating this concern.. 

This stuff gets unwieldy quickly for sure.  I wonder what other factors are influential here.  
personally I have a hard time with how prettier sorts stuff..  but my point isn't to criticize  

I suspect that like with most things, there's always (at least one) exception that makes a judgement call necessary as the world's imperfect and screws are falling out all the time ;)   
(IE: where's the spork go in yer nice three slot utensil drawer?) 
regardless, I don't wanna derail the discussion with my $.03... 

Were that three phase philosophy agreed collectively as `the least-messy-way-forward forever-and-ever`[^1]
do you anticipate this being represented differently?  what makes more sense in your mind when it comes to this? 

The juice has to be worth the squeeze.... an someone's always gonna have some pseudo-legitimate-but-handwavey reason not to hop on the bandwagon...

aside from a reduction in cognitive load by way of providing a clear philosophy guiding attribute modifier sorting...  what benefits would you expect to achieve?  

it feels to me like this is a multidimensional problem that only accepts input on two dimensions.... IOW something's always gonna not really be elegant to express. 

again, just wanted to express some gratitude for the energy you took writing this up so much more elegantly than I would have. 
excited to see others' opinions

[^1]: Until next tuesday, when we invent the  foone .. ( or would it be a sorke?)

**Reply by asimpletune** [reply]
Thank you for your kind words. Let me try to respond to your main points.

> Were that three phase philosophy agreed collectively as the least-messy-way-forward forever-and-eve
do you anticipate this being represented differently?

I'm not 100% sure I understand 'that three phase philosophy', so just understand that I'm assuming you meant the 3 rules that are used to decide sort order that I wrote above. Please correct me if I'm wrong. The 3 rules for CSS classname sorting were:

> 1. user defined classes
> 2. layout first, then decoration
> 3. like with like, with the more specific coming later

So, 'do you anticipate this being represented differently?' No, I don't personally anticipate much debate over the interpretation of these rules. The only one that is subjective is #2, and I think the tailwind people have already made the right decisions regarding that. There's not really a dispute over that, at least not from me.

> aside from a reduction in cognitive load by way of providing a clear philosophy guiding attribute modifier sorting... what benefits would you expect to achieve?

None. There is no benefit to the tailwind prettier plugin than a 'reduction in cognitive load', which is why the class sort order is important. 

Let me take this chance to reiterate that those 3 rules are not my proposal. I took them directly from Tailwind's documentation. The only thing that I suggested is to exclude classes that have modifiers from being exempt by those 3 rules, and the reasons why are best articulated in my post, so I won't repeat them here.

More than likely it is the way it is now for a very good reason, which is that it was just easier for the contributors to implement. The sort order comes from the order the classes are declared in the generated CSS. Ease and simplicity of implementation is a great reason to do something, and I 100% support that. 

Eventually though, the best sort order for the user is what should happen, if there's time and resources to implement it. I don't know if my proposal is the best, but I gave my reasons why I think it is at least an improvement. After examining my points we can determine my mistake, but if I'm not mistaken then we shouldn't shy from the challenge of making improvements when we know there's a better way.

I hope I've answered your questions.

**Comment by KennethHoff** [comment]
I don't really have an opinion, but this is a well written post! 😃

**Comment by laptopmutia** [comment]
I just wish we have official non prettier plugin like https://github.com/heybourn/headwind or https://github.com/avencera/rustywind

it's hard to make the prettier plugin works with template engine like `*.html.erb`

with headwind the class sorter just works seamlessy without any tangled javascript settings and NPM dependencies hell

too bad it's unmaintained and not supporting newer syntax from tailwind 3


so here is my current notes

so there is an emacs lsp-mode https://github.com/merrickluo/lsp-tailwindcss based on https://github.com/tailwindlabs/tailwindcss-intellisense and it use https://github.com/avencera/rustywind that is inspired by https://github.com/heybourn/headwind an non prettier class sorting that is obsolete

problem is prettier hard to setup especially when you use template engine like .erb 

when I use headwind I just install the extension and its just works, 

with prettier I got tangled by the settings and npm dependencies hell

**Reply by rasibn** [reply]
An official rustywind is very much needed for non-js templating syntax or some sort of regex command
